### PR TITLE
ARROW-15661: [Gandiva][C++] Add SHA512 function

### PR DESCRIPTION
**Pull request by Johnnathanalmeida**
This PR was created to implement SHA512 functions in Gandiva side. 

This PR implements the follow signatures:

FunctionSignature{name =hashSHA512, return type =String, param types =[int8]}
FunctionSignature{name =hashSHA512, return type =String, param types =[int16]}
FunctionSignature{name =hashSHA512, return type =String, param types =[int32]}
FunctionSignature{name =hashSHA512, return type =String, param types =[int64]}

FunctionSignature{name =hashSHA512, return type =String, param types =[uint8]}
FunctionSignature{name =hashSHA512, return type =String, param types =[uint16]}
FunctionSignature{name =hashSHA512, return type =String, param types =[uint32]}
FunctionSignature{name =hashSHA512, return type =String, param types =[uint64]}

FunctionSignature{name =hashSHA512, return type =String, param types =[float]}
FunctionSignature{name =hashSHA512, return type =String, param types =[double]}

FunctionSignature{name =hashSHA512, return type =String, param types =[boolean]}

FunctionSignature{name =hashSHA512, return type =String, param types =[date64]}
FunctionSignature{name =hashSHA512, return type =String, param types =[date32]}

FunctionSignature{name =hashSHA512, return type =String, param types =[time32]}
FunctionSignature{name =hashSHA512, return type =String, param types =[timestamp]}

FunctionSignature{name =hashSHA512, return type =String, param types =[string]}

### CODE REVIEW

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 1]
```diff
@@ -1022,6 +1024,73 @@ const char* gdv_mask_first_n_utf8_int32(int64_t context, const char* data,
   return out;
 }
 
+const char* to_hex_binary(int64_t context, const char* text, int32_t text_len,
```
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:1027]: Why is this function added? I dont see any calls to this function

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 2]
```diff
@@ -1022,6 +1024,73 @@ const char* gdv_mask_first_n_utf8_int32(int64_t context, const char* data,
...
+    const char* sha_hash = gandiva::gdv_sha512_hash(context, data, data_len, out_len);
+
+    if (sha_hash == nullptr) {
+      gdv_fn_context_set_error_msg(context,
+                                   "Could not allocate memory for output string");
+      *out_len = 0;
+      return "";
+    }
+
+    return gdv_fn_lower_utf8(context, sha_hash, *out_len, out_len);
```
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:1079]: is this required? This looks like it converting to lower case utf-8 -- this function call will take time to evaluate

File: cpp/src/gandiva/gdv_function_stubs_test.cc [hunk 3]
```diff
@@ -949,4 +950,175 @@ TEST(TestGdvFnStubs, TestMaskLastN) {
   EXPECT_EQ(expected, std::string(result, out_len));
 }
 
+TEST(TestGdvFnStubs, TestMaskHash) {
+  gandiva::ExecutionContext ctx;
+  int64_t ctx_ptr = reinterpret_cast<int64_t>(&ctx);
+  int32_t out_len = 0;
+
+  if (FIPS_mode()) {
```
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:958]: Have you run these tests with FIPS_mode() enabled and disabled?

Comment (Johnnathanalmeida) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:958]: I don't know how to enable this method.

This library even has a method to set FIPS_mode to true, but it didn't work, it always disabled.

So to test the block of code that executes SHA512, I set hard code true in this condition and compared the return of SHA512 with other external sites.

I tested all returns from sha256 and sha512 on external sites to validate:
https://emn178.github.io/online-tools/sha256.html
https://emn178.github.io/online-tools/sha512.html
https://passwordsgenerator.net/sha512-hash-generator/

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 2 above)
Comment (Johnnathanalmeida) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Apparently it is not mandatory.
I removed the call to the lower function and the tests continued to pass without errors.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 4]
```diff
@@ -20,13 +20,13 @@
 #include <utf8proc.h>
 
 #include <boost/crc.hpp>
+#include <iomanip>
```
Comment (anthonylouisbsb) [inline cpp/src/gandiva/gdv_function_stubs.cc:23]: That import is being used?

**Review by anthonylouisbsb — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 4 above)
Comment (Johnnathanalmeida) [reply cpp/src/gandiva/gdv_function_stubs.cc]: That import was removed.

**Review by siddhantrao23 — APPROVED** [review]

**Pull request merged**